src/framework/mpas_io.F

```diff
@@ -5033,6 +5037,149 @@ subroutine MPAS_io_get_att_real1d(handle, attName, attValue, fieldname, precisio
 
    end subroutine MPAS_io_get_att_real1d
 
+   function handle_put_att_pio_redef(handle) result (pio_ierr)
+      implicit none
+      type(MPAS_IO_Handle_type), intent(inout) :: handle
+      integer :: pio_ierr
+
+      call mpas_log_write('Calling PIO_redef')
+      pio_ierr = PIO_redef(handle % pio_file)
+      if (pio_ierr /= PIO_noerr) then
+         io_global_err = pio_ierr
+         return
+      end if
+      call mpas_log_write('Successfully called PIO_redef')
+
+   end function handle_put_att_pio_redef
+
+   function handle_put_att_pio_enddef(handle) result (pio_ierr)
+      implicit none
+      type(MPAS_IO_Handle_type), intent(inout) :: handle
+      integer :: pio_ierr
+
+      call mpas_log_write('Calling PIO_enddef')
+      pio_ierr = PIO_enddef(handle % pio_file)
+      if (pio_ierr /= PIO_noerr) then
+         io_global_err = pio_ierr
+         return
+      end if
+      call mpas_log_write('Successfully called PIO_enddef')
+
+   end function handle_put_att_pio_enddef
+
+   function put_att_0d_generic_pio(handle, varid, attName, attValue, ierr) result(pio_ierr)
+      implicit none
+      type(MPAS_IO_Handle_type), intent(inout) :: handle
+      integer, intent(in) :: varid
+      character(len=*), intent(in) :: attName
+      class(*), intent(in) :: attValue
+      integer, optional :: ierr
+      integer :: pio_ierr
+      character(len=*), parameter :: log_message_prefix = 'Calling PIO_put_att for'
+
+      select type(attValue)
+         type is (integer)
+            call mpas_log_write(log_message_prefix//' integer attribute '//trim(attname))
+            pio_ierr = PIO_put_att(handle % pio_file, varid, attName, attValue)
+         type is (real(kind=R4KIND))
+            pio_ierr = PIO_put_att(handle % pio_file, varid, attName, attValue)
+            call mpas_log_write(log_message_prefix//' real(kind=R4KIND) attribute '//trim(attname))
+         type is (real(kind=R8KIND))
+            call mpas_log_write(log_message_prefix//' real(kind=R8KIND) attribute '//trim(attname))
+            pio_ierr = PIO_put_att(handle % pio_file, varid, attName, attValue)
+         type is (character(len=*))
+            call mpas_log_write(log_message_prefix//' text attribute '//trim(attname))
+            pio_ierr = PIO_put_att(handle % pio_file, varid, attName, attValue)
+      end select
+
+      if (pio_ierr /= PIO_noerr) then
+         io_global_err = pio_ierr
+         if (present(ierr)) ierr = MPAS_IO_ERR_BACKEND
+
+         if (handle % preexisting_file .and. .not. handle % data_mode) then
+            if (handle_put_att_pio_redef(handle) /= PIO_noerr) return
```

Comment:
I think I don't understand what `data_mode` means. I assumed it meant the file is in data mode, in which case you can do data i/o, but can't create new attributes. If so, this logic seems incorrect, because if the file is not in data mode it must be in define mode, so setting the file state to define mode is redundant.

Furthermore, all of the call sites to `put_att_pio` check for `handle%data_mode` and return before doing any work if the file is in data mode. See lines 5376, 5549, and 5719.

What am I missing?

Reply:
It sounds like we think the value of `data_mode`may be incorrect (that is, it isn't set even though the file is in data mode. If that is the case, line 5099 should not check the value of `data_mode`.

Reply:
Yes this seems to be the cause of this problem. I'm looking into all the points you brought above.

Reply:
@jim-p-w I confirmed that this code only works when data mode is false. As you suspected, this means that there is a bug in the code responsible for setting data mode.